add initial serde support for relevant types

vlmutolo opened the pull request:
Resolves #192.

This is still a draft PR. The following items should be addressed before merging.

[ ] No tests are currently implemented. I still have to think about what the best test cases would be. Suggestions are welcome.
[ ] When it was available for the given type, serialization implementations use `unprotected_as_bytes` over the `AsRef<[u8]>` trait impl. Should we prefer using `AsRef` for consistency with types that don't have `unproected_as_bytes`, or is there some reason to prefer `unprotected_as_bytes` for its semantics? I'm leaning toward just using `AsRef` for everything that converts to/from byte slices.
[ ] All the serde-related logic is implemented here in a separate module. I've taken to doing it this way in my own projects to keep the boilerplate out of the modules for each type, but I'm happy to break it up and move the four serde implementations nearer to where the types themselves are defined.
[ ] Another style point: I prefixed most of the types with their module names instead of importing them directly. For example, `kdf::Salt` instead of just `Salt`. I was concerned that other types may implement `Salt` in the future and we may add them to the `serde` module. This is a downside of keeping everything in one module; the names may have to be differentiated further.

vlmutolo commented:
@brycx I have a long weekend, and I'm picking up this draft PR again, along with (hopefully) the CI hardening work. Do you think that it would be better to put the serde implementations next to the types for which they're implemented, or better to put all the serde stuff in one file?

brycx commented:
> Do you think that it would be better to put the serde implementations next to the types for which they're implemented, or better to put all the serde stuff in one file?

I see you've updated the PR description to moving it into `typedefs`. This sounds good to me. Honestly, I believe this is something you're better suited to decide anyway, as you're also implementing it.

vlmutolo commented:
Yeah I'm experimenting with a few different ways of implementing this. I think moving it into typedefs is working out. I was going to ping you when it settles to take a look at decisions involved. 

And yes, before this goes out of draft status, I'll merge master into this and rebase the commits to something more organized.

vlmutolo commented:
Two things that may warrant special attention:

1. I implemented `TryFrom` on a couple types as a thin layer over their "from bytes" implementations. There aren't any tests for the `TryFrom` implementations specifically; it felt like I was just re-testing the code that it calls in the one-line implementations.
2. I implemented most of the `serde`-specific tests in the macro code. Not sure if it's worth testing the higher-level constructions individually. They're all basically the same.

brycx commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ subtle = { version = "^2.2.2", default-features = false }
 zeroize = { version = "1.1.0", default-features = false }
 getrandom = { version = "0.2.0", optional = true }
 ct-codecs = { version = "1.1.1", optional = true }
+serde = { version = "1.0.124", optional = true }
```
There seems to be missing an actual `serde` feature-definition in the `[features]` section of this file.

vlmutolo replied:
I could be misunderstanding this, but I don't think you need to explicitly write down a feature name. Optional dependencies automatically generate features with the same label. Check out [this section of TRPL](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies). Thoughts?

brycx replied:
You are completely right. I didn't know about this. In fact, given that:
> "Note: A feature in the [feature] table cannot use the same name as a dependency."


this seems to be just right.

brycx commented on src/typedefs.rs:
```diff
@@ -85,6 +85,59 @@ macro_rules! impl_normal_debug_trait (($name:ident) => (
+#[cfg(feature = "serde")]
+macro_rules! test_serde_impls (($name:ident, $gen_length:expr) => (
+    #[test]
+    fn test_serde_serialized_equivalence_to_bytes_fn() {
```
We have to have at least one test in the `Tests` workflow compile with the currently-missing `serde` feature, such that these tests are run in CI.

vlmutolo replied:
I think `serde` is the first non-default optional feature added to Orion. Maybe the right way forward is to have an optional feature called `all` or `extra` that will explicitly enable all extra features, including `serde` in this case.

The counterpoint to this is that we don't have any other non-default optional features at the moment, and this could be an example of premature generalization. I could go either way.

For now, for testing, I'll add a new test that's just like normal, except with `serde` enabled. In the future, if we want to add more optional features, maybe we replace that test with something like `cargo test --features all`.

brycx reviewed (changes requested):
This looks great so far! Just two important additions missing, mentioned in comments.

> I implemented TryFrom on a couple types as a thin layer over their "from bytes" implementations. There aren't any tests for the TryFrom implementations specifically; it felt like I was just re-testing the code that it calls in the one-line implementations.

This seems fine to me. `TryFrom` should be implicitly tested through testing `Deserialize` AFAICT.

> I implemented most of the serde-specific tests in the macro code. Not sure if it's worth testing the higher-level constructions individually. They're all basically the same.

The way you tested the `PasswordHash` separately looks good to me. The rest are very much the same. I don't see anything that needs changing here.

brycx reviewed (approved):
LGTM, nice work!

Seems the lints are complaining because `x25519` imports `TryFrom`, but the path to this trait is fully qualified in `typedefs`. If you have the time, it would be nice to have all Action workflows pass before merging. 

If possible, I think it would be fine to exclude this lint in the `typedefs` section of `impl TryFrom`, since we might import `TryFrom` other places in the future, where the `typedefs` macros are also used.

vlmutolo commented:
If you have the time, it would be nice to have all Action workflows pass before merging.Yep! I'll get these lint's passing before I merge. 

brycx commented:
> Ok, good point. I think it would make sense to test serde with alloc enabled, then. At the moment, the deserialization always allocates.

Good idea. Probably, we'll end up adding some `alloc` imports in the `serde` code. `serde` itself also defines an `alloc` features, so this should probably be enabled if orion is also compiled with `alloc` feature enabled.

brycx commented:
Re: 
> "Do you want to do this in scope of this issue, or rather a different one?"

I've created a new issue for this.

vlmutolo commented:
Note that I had to make a small modification in the x25519 code to get the warnings to go away. The only change was that I fully qualified an import (`TryFrom`) instead of using a `use` statement as it was before.

The reason I had to make this change is that something in the x25519 module was importing a macro that fully qualified the `TryFrom` trait, and it was throwing a warning about unnecessary full qualification.
